samples/bluetooth/gatt/CMakeLists.txt

```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_named(samples_bt_gatt)
+
+# We use a number of include/generated/ .h files.
+target_link_libraries(samples_bt_gatt PRIVATE kernel)
```

Comment:
PRIVATE kernel? how is this related to the kernel?

Reply:
> PRIVATE kernel? how is this related to the kernel?

Apparently through libc errno.h, see compilation failures below when you remove this line. Any better way to fix this?

PS: I think it's one of the advantages of using CMake targets: it exposes dependencies. Especially true for west modules (other, similar PRs not submitted yet)

```
Scanning dependencies of target samples_bt_gatt
[  1%] Building C object gatt_bindir/CMakeFiles/samples_bt_gatt.dir/bas.c.obj
In file included from ~/zephyrproject/include/misc/errno_private.h:34,
                 from ~/zephyrproject/lib/libc/minimal/include/errno.h:20,
                 from ~/zephyrproject/samples/bluetooth/gatt/bas.c:14:
tmp/sanity-out/qemu_cortex_m3/samples/bluetooth/peripheral_esp/sample.bluetooth.peripheral_esp/zephyr/include/generated/syscalls/errno_private.h:6:10: fatal error: syscall_list.h: No such file or directory
 #include <syscall_list.h>
          ^~~~~~~~~~~~~~~~
compilation terminated.

[  2%] Building C object gatt_bindir/CMakeFiles/samples_bt_gatt.dir/cts.c.obj
In file included from ~/zephyrproject/include/misc/errno_private.h:34,
                 from ~/zephyrproject/lib/libc/minimal/include/errno.h:20,
                 from ~/zephyrproject/samples/bluetooth/gatt/cts.c:14:
tmp/sanity-out/qemu_cortex_m3/samples/bluetooth/peripheral_esp/sample.bluetooth.peripheral_esp/zephyr/include/generated/syscalls/errno_private.h:6:10: fatal error: syscall_list.h: No such file or directory
 #include <syscall_list.h>
          ^~~~~~~~~~~~~~~~
compilation terminated.

etc.
```

Reply:
Please use `zephyr_library_link_libraries(...)` or `target_link_libraries(${ZEPHYR_CURRENT_LIBRARY}" ...)`. The abstraction in extensions.cmake doesn't guarantee that the target name is the same as the library name. (It happens to work for now, but I'm working on upstreaming a feature that changes that.)

Reply:
> Please use zephyr_library_link_libraries(...)

I couldn't find any `zephyr_` wrapper with PRIVATE, that's why I used `target_link_libraries()`

>  or target_link_libraries(${ZEPHYR_CURRENT_LIBRARY}" ...)

It's really not intuitive to have to use `${ZEPHYR_CURRENT_LIBRARY}` instead of just `my_lib_name` on the line immediately after `zephyr_library_named(my_lib_name)`. Is this going to be the way forward or are there going to be more `zephyr_` wrappers instead? Already discussed in issue #8439 and more generally #8827 (probably better places than here to resume those generic discussions *if needed*)

> The abstraction in extensions.cmake doesn't guarantee that the target name is the same as the library name. (It happens to work for now, but I'm working on upstreaming a feature that changes that.)

Scary. Can you point at more details, some github issue maybe? The documentation of every `target_` command in the CMake book starts with:

> `target_something_something(mytarget ...)`
>
> The named `mytarget` must have been created by a command such as `add_executable(mytarget... )` or `add_library(mytarget ...)` and must not be an ALIAS target.

So unless I'm missing some CMake subtlety, the assumption that the CMake name[*] of a library is identical to the name of its corresponding target seems pervasive across all CMake's documentation and even the `zephyr_` additional layer of indirection didn't seem to complexify that... yet?

[*] not the filename of course

Reply:
> I couldn't find any `zephyr_` wrapper with PRIVATE, that's why I used `target_link_libraries()`

Got it. I missed that, thanks.

> It's really not intuitive to have to use ${ZEPHYR_CURRENT_LIBRARY} instead of just my_lib_name on the line immediately after zephyr_library_named(my_lib_name).

Still, the API of these extensions has never guaranteed that the underlying CMake library's logical target name is the same as the first argument to zephyr_library_named(), so this is not new.

The fact that it happens to work is kind of unfortunate, though; it would have been nice for zephyr_library_named to add some sort of prefix to the library's name as a target. Especially since target names have to be globally unique, some sort of `zephyr-` namespace would have been nice to have from the beginning.

> Is this going to be the way forward or are there going to be more zephyr_ wrappers instead? Already discussed in issue #8439 and more generally #8827 (probably better places than here to resume those generic discussions if needed)

Not sure what you mean by "the way forward": use of `${ZEPHYR_CURRENT_LIBRARY}` allows the native CMake functions to be used, has for some time, and is not going away.

> So unless I'm missing some CMake subtlety, the assumption that the CMake name[*] of a library is identical to the name of its corresponding target seems pervasive across all CMake's documentation and even the zephyr_ additional layer of indirection didn't seem to complexify that... yet?

I agree with your reading of the CMake documentation for its own functions, but that is not what extensions.cmake guarantees or requires:

```
# A Zephyr library can be constructed by the function zephyr_library
# or zephyr_library_named. The constructors create a CMake library
# with a name accessible through the variable ZEPHYR_CURRENT_LIBRARY.
```

(It goes on to say "The variable ZEPHYR_CURRENT_LIBRARY should seldom be needed since
the zephyr libraries have methods that modify the libraries", but I disagree with this on the grounds you've said -- we should be using the CMake functions whenever possible. It's just the target name itself which I want to be careful about here, but there is a variable to dereference to get that name already.)

And finally:

```
# The methods modify the CMake target_* API to reduce boilerplate;
#  PRIVATE is assumed
#  The target is assumed to be ZEPHYR_CURRENT_LIBRARY
```

This is relevant to upstreaming the multi-image feature used downstream by Nordic (https://github.com/zephyrproject-rtos/zephyr/pull/13672). Rework I am doing of that PR relies strongly on the abstraction barrier provided by `${ZEPHYR_CURRENT_LIBRARY}` without introducing any new upstream CMake API requirements.

Reply:
> > `target_link_libraries(samples_bt_gatt PRIVATE kernel)`

@nashif:
> PRIVATE kernel? how is this related to the kernel?

Maybe this was because of
`zephyr_library: missing 'kernel-mode' vs 'app-mode' documentation` #19582